modules/system/defaults/dock.nix

```diff
@@ -134,10 +134,14 @@ in {
       description = ''
         Persistent applications in the dock.
       '';
-      apply = value:
-        if !(isList value)
-        then value
-        else map (app: { tile-data = { file-data = { _CFURLString = app; _CFURLStringType = 0; }; }; }) value;
+      apply =
```

Comment:
I think it would be good if you could change the type to be something more like:

```
[
  { type = "app"; path = "/Applications/Safari.app"; }
  { type = "spacer"; }
  { type = "spacer"; small = true; }
  { type = "folder"; path = "..."; }
]
```

I'll leave the exact types we should have (apps, files, spacers) and options up to you

Reply:
Perhaps a submodule with an enum `type` option, alongside type-specific value options such as `path`?

```nix
types.submodule {
  options = {
    type = mkOption {
      type = types.enum [
        "app"
        # ...
      ];
      default = "app";
      description = "TODO";
    };
    path = mkOption { /* ... */ };
    # ...
  };
}
```

Convenience (and compatibility with the current option) could be achieved using `types.coercedTo`:

```nix
types.coercedTo
  types.str
  (path: {inherit path;})
  (submodule)
```

Reply:
https://nixos.org/manual/nixos/unstable/#sec-option-types-sums

I would use `types.attrTag`